// grapesjs-react/dist/components/GEditor.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var grapesjs_1 = __importDefault(require("grapesjs"));
var grapesjs_blocks_basic_1 = __importDefault(require("grapesjs-blocks-basic"));
var grapesjs_preset_newsletter_1 = __importDefault(require("grapesjs-preset-newsletter"));
// Presets and Plugins
var grapesjs_preset_webpage_1 = __importDefault(require("grapesjs-preset-webpage"));
var react_1 = __importStar(require("react"));
function GEditor(props) {
    var _this = this;
    var _a = props.id, id = _a === void 0 ? 'grapesjs-react-editor' : _a, storageManager = props.storageManager, blockManager = props.blockManager, components = props.components, blocks = props.blocks, webpage = props.webpage, newsletter = props.newsletter;
    var _b = react_1.useState(null), editor = _b[0], setEditor = _b[1];
    react_1.useEffect(function () {
        if (!editor) {
            var plugins = [
                grapesjs_blocks_basic_1.default
            ].concat(props.plugins);
            if (webpage) {
                plugins = plugins.concat([grapesjs_preset_webpage_1.default]);
            }
            if (newsletter) {
                plugins = plugins.concat([grapesjs_preset_newsletter_1.default]);
            }
            var e_1 = grapesjs_1.default.init({
                blockManager: blockManager,
                container: "#" + id,
                fromElement: true,
                plugins: plugins,
                storageManager: storageManager
            });
            var defaultType = e_1.DomComponents.getType('default');
            var defaultModel_1 = defaultType.model;
            var defaultView_1 = defaultType.view;
            components.forEach(function (component) {
                e_1.DomComponents.addType(component.type, {
                    model: defaultModel_1.extend({
                        defaults: Object.assign({}, defaultModel_1.prototype.defaults)
                    }, {
                        isComponent: component.isComponent.bind(_this)
                    }),
                    view: defaultView_1.extend({
                        events: __assign({}, component.events),
                        render: component.render.bind(_this)
                    })
                });
            });
            blocks.forEach(function (block) {
                e_1.BlockManager.add(block.id, block);
            });
            setEditor(e_1);
        }
        else {
            if (document) {
                document.getElementById(id).append(editor.render());
            }
        }
        return function cleanup() {
            if (editor) {
                editor.destroy();
                grapesjs_1.default.editors = grapesjs_1.default.editors.filter(function (e) { return e !== editor; });
            }
        };
    }, []);
    return react_1.default.createElement("div", { id: id });
}
GEditor.defaultProps = {
    blockManager: {},
    blocks: [],
    components: [],
    newsletter: false,
    plugins: [],
    storageManager: {},
    styleManager: {},
    webpage: false,
};
exports.default = GEditor;
window.GrapesJS = grapesjs_1.default;
window.grapesjs = grapesjs_1.default;
